pkg/api/admin/operatorflags.go

```diff
@@ -0,0 +1,63 @@
+package admin
+
+import (
+	"encoding/json"
+	"net/http"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+// Copyright (c) Microsoft Corporation.
+// Licensed under the Apache License 2.0.
+
+type operatorFlagsMergeStrategyStruct struct {
+	OperatorFlagsMergeStrategy string
+	Cluster                    *api.OpenShiftCluster
+}
+
+const (
+	operatorFlagsMergeStrategyDefault string = "merge"
+	operatorFlagsMergeStrategyMerge   string = "merge"
+	operatorFlagsMergeStrategyReset   string = "reset"
+)
+
+// When a cluster is edited via the PATCH Cluster Geneva Action (aka an Admin Update)
+// the flags given are treated according to the provided Update Strategy,
+// provided in operatorFlagsMergeStrategy
+
+// merge (default): The provided cluster flags are laid on top of the cluster’s existing flags.
+// reset: The provided cluster flags are laid on top of the default cluster flags,
+// essentially ‘resetting’ the flags if no new flags are provided.
+func OperatorFlagsMergeStrategy(oc *api.OpenShiftCluster, body []byte) error {
+	payload := operatorFlagsMergeStrategyStruct{
+		OperatorFlagsMergeStrategy: operatorFlagsMergeStrategyDefault,
+		Cluster:                    &api.OpenShiftCluster{},
+	}
+
+	err := json.Unmarshal(body, &payload)
+	if err != nil {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidRequestContent, "", "The request content was invalid and could not be deserialized: %q.", err)
+	}
+
+	// return if payload is empty
+	if properties := &payload.Cluster.Properties; payload.Cluster == nil || properties == nil || payload.Cluster.Properties.OperatorFlags == nil {
+		return nil
+	}
+	// return error if OperatorFlagsMergeStrategy is not merge or reset, default is merge
+	if payload.OperatorFlagsMergeStrategy != operatorFlagsMergeStrategyMerge &&
+		payload.OperatorFlagsMergeStrategy != operatorFlagsMergeStrategyReset {
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeInvalidParameter, "", "invalid operatorFlagsMergeStrategy '%s', can only be 'merge' or 'reset'", payload.OperatorFlagsMergeStrategy)
+	}
+	// return nil, if OperatorFlagsMergeStrategy is merge and payload has not operatorFlags
+	// return operatorFlags of payload, if OperatorFlagsMergeStrategy is merge and payload has operatorFlags
+	// return defaultOperatorFlags, if OperatorFlagsMergeStrategy is reset and payload has not operatorFlags
+	// return defaultOperatorFlags + operatorFlags of payload, if OperatorFlagsMergeStrategy is reset and payload has operatorFlags
```

Comment:
These three lines are a bit confusing. The code should be returning 'error' but these lines say it will return operatorFlags or defaultOperatorFlags. I believe these comments should be changed to mention the code is modifying the  document passed in.

I am also looking for the exact place, where the flags are merged. I cannot find it.

Reply:
It's a bit silly, but it's done in https://github.com/Azure/ARO-RP/blob/614efbb55f6c7b18788cc6773a953fe3a918701f/pkg/frontend/openshiftcluster_putorpatch.go#L158 , because the json decoding automatically merges maps.